pyctcdecode/alphabet.py

```diff
@@ -1,91 +1,111 @@
 # Copyright 2021-present Kensho Technologies, LLC.
+from __future__ import division
+
 import logging
-from typing import List, Optional
+import re
+from typing import Collection, List
+
 
+BPE_TOKEN = "▁"  # representation of token boundary in BPE alphabet # nosec
+UNK_TOKEN = "⁇"  # representation of special UNK token in regular alphabet # nosec
+UNK_BPE_TOKEN = "▁⁇▁"  # representation of special UNK token in BPE alphabet # nosec
 
-BPE_CHAR = "▁"  # character used for token boundary if BPE is used
-UNK_BPE_CHAR = "▁⁇▁"  # representation of unknown character in BPE
+SPECIAL_TOKEN_PTN = re.compile(r"^[<\[].+[>\]]$")
+BLANK_TOKEN_PTN = re.compile(r"^[<\[]pad[>\]]$", flags=re.IGNORECASE)
+UNK_TOKEN_PTN = re.compile(r"^[<\[]unk[>\]]$", flags=re.IGNORECASE)
 
 logger = logging.getLogger(__name__)
 
 
-def _get_ctc_index(label_list: List[str]) -> int:
-    """Get index of ctc blank character in alphabet."""
-    return len(label_list) - 1 if label_list[-1] == "" else -1
-
-
-def _normalize_alphabet(label_list: List[str], ctc_token_idx: Optional[int] = None) -> List[str]:
-    """Normalize alphabet for non-bpe decoder."""
-    if any([len(c) > 1 for c in label_list]):
-        raise ValueError("For non-bpe alphabet only length 1 entries and blank token are allowed.")
-    if ctc_token_idx is None:
-        ctc_token_idx = _get_ctc_index(label_list)
-    clean_labels = label_list[:]
-    # check for space token
-    if " " not in clean_labels:
-        raise ValueError("Space token ' ' missing from vocabulary.")
-    # specify ctc blank token
-    if ctc_token_idx == -1:
-        clean_labels.append("")
+def _check_if_bpe(labels: List[str]) -> bool:
+    """Check if input alphabet is BPE or not."""
+    is_bpe = any([s.startswith("##") for s in labels]) or any(
+        [s.startswith(BPE_TOKEN) for s in labels]
+    )
+    if is_bpe:
+        logger.info("Alphabet determined to be of BPE style.")
     else:
-        clean_labels[ctc_token_idx] = ""
-    return clean_labels
+        logger.info("Alphabet determined to be of regular style.")
+    return is_bpe
+
+
+def _normalize_regular_alphabet(labels: List[str]) -> List[str]:
+    """Normalize non-bpe labels to alphabet for decoder."""
+    normalized_labels = labels[:]
+    # substitute space characters
+    if "|" in normalized_labels and " " not in normalized_labels:
+        logger.info("Found '|' in vocabulary but not ' ', doing substitution.")
+        normalized_labels[normalized_labels.index("|")] = " "
+    # substituted ctc blank char
+    for n, label in enumerate(normalized_labels):
+        if BLANK_TOKEN_PTN.match(label):
+            logger.info("Found %s in vocabulary, substituting with %s.", label, "")
```

Comment:
`("Found %s in vocabulary which was interpreted as a CTC blank token, substituting with the empty string '' ", label)`